pydataproc/dataproc.py

```diff
@@ -95,6 +99,119 @@ def list_clusters(self, minimal=True):
+            cluster_name_filter = cluster_name
+
+        if count > self.MAX_JOBS:
+            log.info('count of {} too high, limiting to {} jobs...'.format(count, self.MAX_JOBS))
+            count = self.MAX_JOBS
+
+        try:
+            result = self.dataproc.projects().regions().jobs().list(
+                projectId=self.project,
+                region=self.region,
+                filter=filter,
+                clusterName=cluster_name_filter
+            ).execute()
+        except HttpError as e:
+            if e.resp['status'] == '404':
+                raise Exception("'{}' is not a valid cluster".format(cluster_name))
+            raise e
+
+        if 'jobs' not in result:
+            return {}
+
+        # TODO investigate using generators here
+        page_token = result['nextPageToken']
+        while len(result['jobs']) < count:
+            moar = self.dataproc.projects().regions().jobs().list(
+                projectId=self.project,
+                region=self.region,
+                filter=filter,
+                clusterName=cluster_name_filter,
+                pageToken=page_token
+            ).execute()
+            if len(moar['jobs']) == 0:
+                break
+            result['jobs'].extend(moar['jobs'])
+            if 'nextPageToken' not in page_token:
+                break
+            page_token = moar['nextPageToken']
+
+        result['jobs'] = result['jobs'][:count]
+
+        if minimal:
+            return {j['reference']['jobId']: j['status']['state'] for j in result.get('jobs', [])}
+        return {j['reference']['jobId']: j for j in result.get('jobs', [])}
+
+    def job_info(self, job_id):
+        """
+        Returns the full configuration information associated with a given
+        job. If the job does not exist, returns None.
+
+        :param job_id: The ID of the job to check
+        :return: dict of job information, or None if no such cluster
+        """
+        assert job_id
+
+        try:
+            return self.dataproc.projects().regions().jobs().get(
+                projectId=self.project,
+                region=self.region,
+                jobId=job_id
+            ).execute()
+        except HttpError as e:
+            if e.resp['status'] == '404':
+                return None
+            raise e
+
+    def stream_job_logs(self, job_id):
+        """
+        Streams the job logs to stdout, using the 'gcloud dataproc jobs wait'
+        command and subprocess.
+
+        :param job_id: string, ID of job to stream logs for
+        :return: None
+        """
+        print('\nJOB LOGS (job ID: {}):\n--------------------------\n'.format(job_id))
+        # piping stdout to PIPE ensures that the job configuration isn't output
+        # when the job completes
+        # This isn't yet supported by the DataProc API/Python lib, so must be done
+        # using subprocess and the gcloud CLI tools
+        subprocess.call(
+            [
+                "gcloud",
```

Comment:
single string and then split()? It's a lot cleaner and readable, IMO.

Reply:
Sure. I did lift this straight from the original implementation, which you also reviewed :D

Reply:
I am allowed to change my mind.